File: packages/python-sdk/src/neon3_sdk/routing.py
"""Intent routing and drag/drop abstractions (Stage 005).

``IntentRouter`` turns any business semantic/drop event into a typed handler
call and a resulting publication. It matches an exact intent first, then a
registered prefix (``domain.item.*``), then an optional default handler; an
unmatched intent raises a structured :class:`UnsupportedIntentError` instead
of being silently dropped. Handlers may be sync or async and may return
``None`` (no state change), a ``UiHostPublication`` dict, or a store
transaction result.

Drag/drop is declarative and generic: a :class:`DragSource` binds a stable
node key to a payload factory and :class:`DropTarget` records which payload
*kinds* it accepts and which intent a completed drop dispatches. The router
resolves a raw ``UiProgramDragDropEvent`` into a :class:`DropEvent` carrying
the SDK-level business payload, then routes on the target's intent. Nothing
here knows about inventory, equipment or recipes — those live in examples.
"""

from __future__ import annotations

import inspect
from dataclasses import dataclass, field
from typing import Any, Awaitable, Callable, Iterable

from .errors import DropRejectedError, UnknownTargetError, UnsupportedIntentError
from .models import DropEvent, IntentEvent

HandlerResult = Any
IntentHandler = Callable[[Any], HandlerResult | Awaitable[HandlerResult]]


@dataclass(frozen=True)
class DragSource:
    """A stable node key plus the business payload a completed drag yields."""

    key: str
    payload_for: Callable[[Any], dict[str, Any]] = field(default=lambda item: {"key": item} if not isinstance(item, dict) else dict(item))
    kind_of: Callable[[Any], str] = field(default=lambda item: str(item.get("kind") if isinstance(item, dict) else type(item).__name__))

    def payload(self, item: Any) -> dict[str, Any]:
        return dict(self.payload_for(item))

    def kind(self, item: Any) -> str:
        return self.kind_of(item)


@dataclass(frozen=True)
class DropTarget:
    """A stable node key that accepts certain payload kinds and emits ``intent``."""

    key: str
    intent: str
    accepts: tuple[str, ...] = ()

    def accepts_kind(self, kind: str) -> bool:
        return not self.accepts or kind in self.accepts


class IntentRouter:
    """Exact/prefix/default intent dispatch with typed rejections."""

    def __init__(self) -> None:
        self._exact: dict[str, IntentHandler] = {}
        self._prefix: list[tuple[str, IntentHandler]] = []
        self._default: IntentHandler | None = None
        self._drag_sources: dict[str, DragSource] = {}
        self._drop_targets: dict[str, DropTarget] = {}
        self._catalog: dict[str, Any] = {}  # stable node key -> domain item

    # -- registration -------------------------------------------------------

    def on(self, intent: str, handler: IntentHandler | None = None):
        """Register a handler for an exact intent (or ``prefix.*``).

        Usable directly (``router.on("x", fn)``) or as a decorator
        (``@router.on("x")``).
        """
        def _register(function: IntentHandler) -> IntentHandler:
            if intent.endswith(".*"):
                self._prefix.append((intent[:-1], function))
            elif intent == "*":
                self._default = function
            else:
                self._exact[intent] = function
            return function

        if handler is None:
            return _register
        return _register(handler)

    def default(self, handler: IntentHandler):
        self._default = handler
        return handler

    def drag_source(self, key: str, payload: Callable[[Any], dict[str, Any]] | None = None, *, kind_of: Callable[[Any], str] | None = None) -> DragSource:
        source = DragSource(key=key, payload_for=payload or DragSource.payload_for, kind_of=kind_of or DragSource.kind_of)
        self._drag_sources[key] = source
        return source

    def drop_target(self, key: str, intent: str, accepts: Iterable[str] | None = None) -> DropTarget:
        target = DropTarget(key=key, intent=intent, accepts=tuple(accepts or ()))
        self._drop_targets[key] = target
        return target

    def catalog(self, items: dict[str, Any]) -> None:
        """Bind stable node keys to domain items so drops can resolve payload."""
        self._catalog = dict(items)

    def register_catalog(self, key_of: Callable[[Any], str], items: Iterable[Any]) -> None:
        self._catalog = {key_of(item): item for item in items}

    # -- lookup -------------------------------------------------------------

    def has_intent(self, intent: str) -> bool:
        if intent in self._exact or self._default is not None:
            return True
        return any(intent.startswith(prefix) for prefix, _ in self._prefix)

    def handler_for(self, intent: str) -> IntentHandler:
        if intent in self._exact:
            return self._exact[intent]
        for prefix, handler in self._prefix:
            if intent.startswith(prefix):
                return handler
        if self._default is not None:
            return self._default
        raise UnsupportedIntentError(intent)

    # -- dispatch -----------------------------------------------------------

    async def dispatch(self, event: IntentEvent | DropEvent | dict[str, Any]) -> HandlerResult:
        """Route one semantic or drop event to its handler.

        Accepts an :class:`IntentEvent`, a :class:`DropEvent`, or a raw
        ``UiHostInbound`` envelope. Returns whatever the handler returns
        (``None``, a publication dict, or a store result). Raises
        :class:`UnsupportedIntentError`, :class:`UnknownTargetError`, or
        :class:`DropRejectedError` for the three distinguishable outcomes.
        """
        if isinstance(event, dict):
            event = self.resolve_inbound(event)
        if isinstance(event, DropEvent):
            handler = self.handler_for(event.intent or "")
        else:
            handler = self.handler_for(event.intent)
        result = handler(event)
        if inspect.isawaitable(result):
            result = await result
        return result

    def resolve_inbound(self, inbound: dict[str, Any]) -> IntentEvent | DropEvent:
        """Turn a raw host inbound envelope into a typed SDK event.

        A ``drag_drop`` resolves its SDK payload through the registered drag
        source and catalog, verifies the drop target exists (else
        :class:`UnknownTargetError`) and that it accepts the payload kind
        (else :class:`DropRejectedError`, domain state untouched).
        """
        kind = inbound.get("kind")
        event = inbound.get("event") or {}
        if kind == "semantic_intent":
            return IntentEvent.from_inbound(event)
        if kind == "drag_drop":
            wire = event.get("payload") or {}
            source_key = str(wire.get("source_key", ""))
            target_key = str(wire.get("target_key", ""))
            if target_key not in self._drop_targets:
                raise UnknownTargetError(target_key, kind="drop")
            if source_key not in self._drag_sources:
                raise UnknownTargetError(source_key, kind="drag")
            target = self._drop_targets[target_key]
            item = self._catalog.get(source_key)
            source = self._drag_sources[source_key]
            payload = source.payload(item) if item is not None else {}
            drop_kind = str(payload.get("kind", source.kind(item)))
            if not target.accepts_kind(drop_kind):
                raise DropRejectedError(
                    f"drop target {target_key!r} does not accept kind {drop_kind!r}",
                    source_key=source_key,
                    target_key=target_key,
                    accepted=target.accepts,
                )
            # The runtime enforces drop_record.intent == event.intent, so the
            # target's declared intent is authoritative when the envelope is
            # terse (SDK-built test fixtures) and agrees when it is not.
            resolved = {**event, "intent": event.get("intent") or target.intent}
            return DropEvent.from_inbound(resolved, payload=payload)
        raise UnsupportedIntentError(f"<unknown inbound kind: {kind}>")

    # -- introspection ------------------------------------------------------

    @property
    def intents(self) -> tuple[str, ...]:
        return tuple(sorted(self._exact))

    @property
    def drop_intents(self) -> dict[str, str]:
        return {key: target.intent for key, target in self._drop_targets.items()}

File: packages/node-sdk/src/routing.ts
/**
 * Intent routing and drag/drop abstractions (Stage 005).
 *
 * `IntentRouter` turns any business semantic/drop event into a typed handler
 * call. It matches an exact intent first, then a registered prefix
 * (`domain.item.*`), then an optional default handler; an unmatched intent
 * throws a structured UnsupportedIntentError instead of being silently
 * dropped. Handlers may be sync or async.
 *
 * Drag/drop is declarative and generic. A `DragSource` binds a stable node key
 * to a payload factory; a `DropTarget` records which payload kinds it accepts
 * and which intent a completed drop dispatches. Mirrors `neon3_sdk/routing.py`
 * — nothing here knows about inventory, equipment or recipes.
 */

import { DropRejectedError, UnknownTargetError, UnsupportedIntentError } from "./errors.js";
import { DropEvent, IntentEvent } from "./protocol.js";

export interface DragSourceSpec {
  key: string;
  payloadFor: (item: unknown) => Record<string, unknown>;
  kindOf: (item: any) => string;
}

export interface DropTargetSpec {
  key: string;
  intent: string;
  accepts: string[];
}

export type IntentHandler = (event: IntentEvent | DropEvent) => unknown | Promise<unknown>;

const defaultPayloadFor = (item: any): Record<string, unknown> =>
  item && typeof item === "object" ? { ...item } : { key: item };
const defaultKindOf = (item: any): string =>
  item && typeof item === "object" && "kind" in item ? String(item.kind) : typeof item;

export class DragSource implements DragSourceSpec {
  constructor(readonly key: string, readonly payloadFor: (item: any) => Record<string, unknown> = defaultPayloadFor, readonly kindOf: (item: any) => string = defaultKindOf) {}
  payload(item: any): Record<string, unknown> { return { ...this.payloadFor(item) }; }
  kind(item: any): string { return this.kindOf(item); }
}

export class DropTarget implements DropTargetSpec {
  constructor(readonly key: string, readonly intent: string, readonly accepts: string[] = []) {}
  acceptsKind(kind: string): boolean { return this.accepts.length === 0 || this.accepts.includes(kind); }
}

export class IntentRouter {
  private readonly exact = new Map<string, IntentHandler>();
  private readonly prefixes: Array<[string, IntentHandler]> = [];
  private defaultHandler: IntentHandler | null = null;
  private readonly dragSources = new Map<string, DragSource>();
  private readonly dropTargets = new Map<string, DropTarget>();
  private catalog = new Map<string, any>();
  private readonly bindings: Array<any> = [];

  /** Register a handler for an exact intent or `prefix.*` (or `*` default). */
  on(intent: string, handler?: IntentHandler): IntentHandler | ((handler: IntentHandler) => IntentHandler) {
    const register = (function_: IntentHandler): IntentHandler => {
      if (intent.endsWith(".*")) this.prefixes.push([intent.slice(0, -1), function_]);
      else if (intent === "*") this.defaultHandler = function_;
      else this.exact.set(intent, function_);
      return function_;
    };
    return handler === undefined ? register : register(handler);
  }

  default(handler: IntentHandler): IntentHandler {
    this.defaultHandler = handler;
    return handler;
  }

  dragSource(key: string, options: { payload?: (item: any) => Record<string, unknown>; kindOf?: (item: any) => string } = {}): DragSource {
    const source = new DragSource(key, options.payload ?? defaultPayloadFor, options.kindOf ?? defaultKindOf);
    this.dragSources.set(key, source);
    return source;
  }

  dropTarget(key: string, intent: string, accepts?: string[]): DropTarget {
    const target = new DropTarget(key, intent, accepts ?? []);
    this.dropTargets.set(key, target);
    return target;
  }

  /** Bind stable node keys to domain items so drops can resolve payload. */
  setCatalog(items: Iterable<any>, keyOf: (item: any) => string): void {
    this.catalog = new Map();
    for (const item of items) this.catalog.set(keyOf(item), item);
  }

  setCatalogMap(items: Record<string, any>): void {
    this.catalog = new Map(Object.entries(items));
  }

  addBinding(binding: any): void { this.bindings.push(binding); }

  hasIntent(intent: string): boolean {
    if (this.exact.has(intent) || this.defaultHandler) return true;
    return this.prefixes.some(([prefix]) => intent.startsWith(prefix));
  }

  handlerFor(intent: string): IntentHandler {
    const exact = this.exact.get(intent);
    if (exact) return exact;
    for (const [prefix, handler] of this.prefixes) if (intent.startsWith(prefix)) return handler;
    if (this.defaultHandler) return this.defaultHandler;
    throw new UnsupportedIntentError(intent);
  }

  /**
   * Route one semantic or drop event to its handler. Accepts an IntentEvent,
   * a DropEvent, or a raw host inbound envelope. Returns the handler result
   * (a promise when the handler is async). Throws UnsupportedIntentError,
   * UnknownTargetError, or DropRejectedError for the three outcomes.
   */
  async dispatch(event: IntentEvent | DropEvent | Record<string, unknown>): Promise<unknown> {
    const resolved = event instanceof IntentEvent || event instanceof DropEvent ? event : this.resolveInbound(event as Record<string, unknown>);
    const intent = resolved instanceof DropEvent ? resolved.intent ?? "" : resolved.intent;
    const handler = this.handlerFor(intent);
    const result = handler(resolved);
    return result instanceof Promise ? await result : result;
  }

  /** Turn a raw host inbound envelope into a typed SDK event. */
  resolveInbound(inbound: Record<string, unknown>): IntentEvent | DropEvent {
    const kind = inbound.kind;
    const event = (inbound.event ?? {}) as Record<string, any>;
    if (kind === "semantic_intent") return IntentEvent.fromInbound(event);
    if (kind === "drag_drop") {
      const wire = (event.payload ?? {}) as Record<string, unknown>;
      const sourceKey = String(wire.source_key ?? "");
      const targetKey = String(wire.target_key ?? "");
      const target = this.dropTargets.get(targetKey);
      if (!target) throw new UnknownTargetError(targetKey, "drop");
      let payload: Record<string, unknown>;
      let kind2: string;
      const source = this.dragSources.get(sourceKey);
      if (source) {
        const item = this.catalog.get(sourceKey);
        payload = item !== undefined ? source.payload(item) : {};
        kind2 = String(payload.kind ?? source.kind(item));
      } else {
        const binding = this.bindings.find((candidate) => candidate.keyForNode(sourceKey) !== null);
        if (!binding) throw new UnknownTargetError(sourceKey, "drag");
        const itemKey = binding.keyForNode(sourceKey) as string;
        const item = binding.source.get(itemKey);
        const [, payloadFor, kindFor] = binding.dragSourceSpec();
        payload = payloadFor(item);
        kind2 = String(payload.kind ?? kindFor(item));
      }
      if (!target.acceptsKind(kind2)) {
        throw new DropRejectedError(`drop target '${targetKey}' does not accept kind '${kind2}'`, {
          sourceKey,
          targetKey,
          accepted: target.accepts,
        });
      }
      // The runtime enforces drop_record.intent == event.intent, so the
      // target's declared intent is authoritative when the envelope is terse.
      return DropEvent.fromInbound({ ...event, intent: event.intent ?? target.intent }, payload);
    }
    throw new UnsupportedIntentError(`<unknown inbound kind: ${String(kind)}>`);
  }

  get intents(): string[] { return [...this.exact.keys()].sort(); }

  get dropIntents(): Record<string, string> {
    const map: Record<string, string> = {};
    for (const [key, target] of this.dropTargets) map[key] = target.intent;
    return map;
  }
}
